aws-s3-transfer-manager/src/io/stream.rs

```diff
@@ -94,13 +133,110 @@ pub(super) enum RawInputStream {
     Buf(Bytes),
     /// File based input
     Fs(PathBody),
+    /// User provided custom stream
+    Dyn(BoxStream),
+}
+
+/// The context of an input stream.
+#[derive(Debug)]
+pub struct StreamContext {
+    part_size: usize,
+}
+
+impl StreamContext {
+    pub(super) fn new(part_size: usize) -> Self {
+        Self { part_size }
+    }
+
+    /// The part size to use when yielding parts.
+    /// NOTE: this _may_ differ from the configured part size (e.g. if the target part size would
+    /// result in exceeding the maximum number of parts allowed).
+    pub fn part_size(&self) -> usize {
+        self.part_size
+    }
+
+    // TODO - eventually make the ability to allocate a buffer public after carefully review of the `Buffer` API.
+    /// Request a new buffer to fill
+    pub(crate) fn new_buffer(&self, capacity: usize) -> Buffer {
+        // TODO - replace allocation with memory pool
+        Buffer::new(capacity)
+    }
+}
+
+/// Contents and (optional) metadata for a single part of a [multipart upload].
+///
+/// [multipart upload]: https://docs.aws.amazon.com/AmazonS3/latest/userguide/mpuoverview.html
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct PartData {
+    // 1-indexed
+    pub(crate) part_number: u64,
+    pub(crate) data: Bytes,
+}
+
+impl PartData {
+    /// Create a new part
+    pub(super) fn new(part_number: u64, data: impl Into<Bytes>) -> Self {
+        Self {
+            part_number,
+            data: data.into(),
+        }
+    }
+}
+
+/// Trait representing a stream of object parts (streaming body).
+///
+/// Individual parts are streamed via the `poll_part` function, which asynchronously yields
+/// instances of `PartData`. When `Poll::Ready(None)` is returned the stream is assumed to have
+/// reached EOF and is finished.
+///
+/// The `size_hint` function provides insight into the total number of bytes that will be streamed.
+pub trait PartStream {
+    /// Attempt to pull the next part from the stream.
+    ///
+    /// The `stream_cx` will have the part size that should be utilized. Implementations should be
+    /// careful to only yield full parts for every part except the last one, which _may_ be less
+    /// than the full part size.
+    fn poll_part(
```

Comment:
> I don't think we would today hit a deadlock.

Cool, good to know. But I expect our internal scheduling to undergo a lot of changes in the next few months. This won't be an issue until we have some kind of memory limiter trying to be smart about the total number of parts in memory across all open uploads. 

It's just a design requirement we need to keep in mind as we build this stuff out. And definitely we should have a test to make sure future changes don't break it. [Here's the test](https://github.com/awslabs/aws-c-s3/blob/5d8d4205e7de4e152bf26bb27d86f3acfa8cd5d2/tests/s3_many_async_uploads_without_data_test.c#L12) in aws-c-s3

> Backpressure has to happen somewhere. If aws-c-s3 returns PENDING what does the user/mountpoint do?

with aws-c-s3's poll_write() function, it must be called sequentially (you can't have multiple threads calling it at the same time), which works fine for mountpoint because file writes are also sequential. It can impose backpressure by returning PENDING if the object already has 1+ parts submitted, but not yet uploaded.

Reply:
Agreed we need a test for this. I suspect we can adapt a write solution with `poll_read` via channels.

e.g.

```rs
struct ChanPartStream {
     rx: Receiver<Option<Result<PartData, io::Error>>
}
struct Sender {
    tx: Sender<Option<Result<PartData, io::Error>>
}

impl Sender {
    async fn send(part: Result<PartData, io::Errror>) { ... }

    // returns error/PENDING on full channel buffer
    fn try_send(part: Result<PartData, io::Error>) -> Result<(), TrySendError> { ... }
}

fn new_chan_stream() -> (InputStream, Sender) { ... }

impl PartStream for ChanPartStream {
     fn poll_stream(...) { 
          self.rx.poll_recv(cx)
     }
}

```

Reply:
@graebm FYI added a test for this in https://github.com/awslabs/aws-s3-transfer-manager-rs/pull/70/commits/0fdbd651fa1eea37fbd27a51f50eb5ec158a27b1